binder/src/main/java/io/grpc/binder/internal/ServiceBinding.java

```diff
@@ -117,29 +125,65 @@ private void notifyUnbound(Status reason) {
   public synchronized void bind() {
     if (state == State.NOT_BINDING) {
       state = State.BINDING;
-      Status bindResult = bindInternal(sourceContext, bindIntent, this, bindFlags);
+      Status bindResult =
+          bindInternal(channelCredentials, bindIntent, this, bindFlags, targetUserHandle);
       if (!bindResult.isOk()) {
-        handleBindServiceFailure(sourceContext, this);
+        handleBindServiceFailure(channelCredentials.getSourceContext(), this);
         state = State.UNBOUND;
         mainThreadExecutor.execute(() -> notifyUnbound(bindResult));
       }
     }
   }
 
   private static Status bindInternal(
-      Context context, Intent bindIntent, ServiceConnection conn, int flags) {
+      BinderChannelCredentials channelCredentials,
+      Intent bindIntent,
+      ServiceConnection conn,
+      int flags,
+      @Nullable UserHandle targetUserHandle) {
+    Context context = channelCredentials.getSourceContext();
+    String methodName = "bindService";
     try {
-      if (!context.bindService(bindIntent, conn, flags)) {
-        return Status.UNIMPLEMENTED.withDescription(
-            "bindService(" + bindIntent + ") returned false");
+      if (targetUserHandle == null) {
```

Comment:
This new version is hard to read because it's now long and gets deeply nested (go/tott/553)

It appears that the only real change here is a choice of bindService() implementation. Can you replace conditionals with polymorphism (https://refactoring.guru/replace-conditional-with-polymorphism)

Reply:
I don't think polymorphism can be applied easily here. This method is a static method.  I did refactor this part a little bit to make the nested if/else to a switch statement. Feel free to let me know your thoughts.